Batteries/CodeAction/Match.lean

```diff
@@ -0,0 +1,192 @@
+def myfun2 (n : Nat) : Nat :=
+  match n with
+  | 0 => _
+  | n + 1 => _
+```
+Also has support for multiple discriminants, e.g.
+```
+def myfun3 (o : Option Bool) (m : Nat) : Nat :=
+  match o, m with
+```
+can be expanded into
+```
+def myfun3 (o : Option Bool) (m : Nat) : Nat :=
+  match o, m with
+  | .none, 0 => _
+  | .none, n_2 + 1 => _
+  | .some val_1, 0 => _
+  | .some val_1, n_2 + 1 => _
+```
+-/
+@[command_code_action] --I couldn't make this work with '@[command_code_action Parser.Term.match]':
+                       --It never fires. So i filter it myself in Step 1.
+def matchExpand : CommandCodeAction := fun CodeActionParams snap ctx node => do
+  -- 1. Find ALL ofTermInfo Info nodes that are of kind `Term.match`
+  let allMatchInfos := findAllInfos isMatchTerm node
+
+  /- 2. Filter these candidates within the `RequestM` monad based on the cursor being in the
+  header lines of these matches. -/
+  let doc ← readDoc
+  let relevantMatchInfos ← allMatchInfos.filterM fun matchInfo => do
+    let some headerRangeRaw := getMatchHeaderRange? matchInfo.stx | return false
+    let headerRangeLsp := doc.meta.text.utf8RangeToLspRange headerRangeRaw
+
+    let cursorRangeLsp := CodeActionParams.range
+    -- check if the cursor range is contained in the header range
+    return (cursorRangeLsp.start ≥ headerRangeLsp.start && cursorRangeLsp.end ≤ headerRangeLsp.end)
+
+  /- 3. Pick the first (and mostly only) candidate. There might sometimes be more,
+  since some things are just contained multiple times in 'node'. -/
+  let some matchInfo := relevantMatchInfos[0]? | return #[]
+  let some headerRangeRaw := getMatchHeaderRange? matchInfo.stx | return #[]
+
+  /- Isolate the array of match-discriminants -/
+  let discrs ← match matchInfo.stx with
+  | `(term| match
+    $[(generalizing := $generalizingVal)]?
+    $[(motive := $motiveVal)]?
+    $[$discrs:matchDiscr],*
+    with $_) => pure discrs
+  | _ => return #[]
+
+  /- Reduce discrs to the array of match-discriminants-terms (i.e. "[n1, n2]" in "match n2,n2"). -/
+  let some discrTerms := discrs.mapM (fun discr =>
+    match discr with
+    | `(matchDiscr| $t: term) => some t
+    | `(matchDiscr| $_:ident : $t: term) => some t
+    | _ => none
+    ) | return #[]
+
+  -- Get a Bool, that tells us if "with" is already typed in:
+  let withPresent :=
+    (matchInfo.stx.getArgs.find? (fun s => s.isAtom && s.getAtomVal == "with")).isSome
+
+  /- Construct a list containing for each discriminant its list of constructornames: -/
+  let mut constructors : List (List Name) := []
+  for discrTerm in discrTerms do
+    let some (info : TermInfo) := findTermInfo? node discrTerm | return #[]
+    let ty ← info.runMetaM ctx (Lean.Meta.inferType info.expr)
+    let .const name _ := (← info.runMetaM ctx (whnf ty)).getAppFn | return #[]
+    -- Find the inductive constructors of e:
+    let some (.inductInfo val) := snap.env.find? name | return #[]
+    constructors := constructors.concat (val.ctors)
```

Comment:
`val.ctors :: constructors` would be more efficient and the constructors list would already be reversed for the usage below.